power_grid_model_c/power_grid_model/include/power_grid_model/sparse_mapping.hpp

```diff
@@ -50,35 +50,164 @@ struct SparseMapping {
         sparse_mapping.reorder[--counter[it_entry->second]] = it_entry->first;
     }
+    // the first is the index before reordering of load_gen
+    // it entry second is the index of the bus
+    // reorder at the end contains for each new index of the load gen the old index of the load gen in a linear way
     assert(sparse_mapping.indptr[0] == 0);
     assert(sparse_mapping.indptr.back() == n_A);
     return sparse_mapping;
 }
 
-} // namespace power_grid_model
+struct DenseMapping {
+    IdxVector indvector;
+    IdxVector reorder;
+};
+
+inline DenseMapping build_dense_mapping(IdxVector const& idx_B_in_A, Idx const n_B) {
+    auto const n_A = static_cast<Idx>(idx_B_in_A.size());
+    auto const r = std::max_element(idx_B_in_A.begin(), idx_B_in_A.end())[0];
+    // for counting sort nr_of_operations = 1 + 3*n_A + 2*n_B + 2*n_A + 2*n_B - 1 + 1 + n_A + 1 + 14 * n_A
+    // nr_of_operations = 20*n_A + 4*n_B + 2
+    auto const c = 20 * n_A + 4 * n_B + 2;
+    // for n log n sort nr of operations = 1 + 7*n_A + n_A*log(n_A)?
+    using DenseEntry = std::pair<Idx, Idx>;
+    DenseMapping dense_mapping;
+
+    if (n_A * log(n_A) < c * r) {
+        // 1 operation?
+        std::vector<DenseEntry> mapping_to_from;
+
+        // n_A operations
+        auto index_range = std::ranges::views::iota(0, static_cast<Idx>(idx_B_in_A.size()));
+
+        // n_A * 2 operations?
+        std::ranges::transform(idx_B_in_A, index_range, std::back_inserter(mapping_to_from),
+                               [](Idx value, Idx orig_idx) {
+                                   return std::pair{value, orig_idx};
+                               });
+        // n_A log n_A?
+        std::ranges::sort(mapping_to_from); // runs in O(n log n)
+
+        // n_A * 2 operations?
+        std::ranges::transform(mapping_to_from, std::back_inserter(dense_mapping.indvector),
+                               [](DenseEntry const& to_from) { return to_from.first; });
+        // n_A * 2 operations?
+        std::ranges::transform(mapping_to_from, std::back_inserter(dense_mapping.reorder),
+                               [](DenseEntry const& to_from) { return to_from.second; });
+    } else {
+        // 1
+        std::vector<DenseEntry> entries(n_A);
+
+        // n_A, since the transform operation will be performed n_A times and it only does one operation - returning the
+        // dense entry
+        std::transform(idx_B_in_A.cbegin(), idx_B_in_A.cend(), IdxCount{0}, entries.begin(), [](Idx j_B, Idx i_A) {
+            return DenseEntry{i_A, j_B};
+        });
+
+        // n_A or can we say that resize is constant 1?
+        dense_mapping.indvector.resize(n_A);
+        // n_A or can we say that resize is constant 1?
+        dense_mapping.reorder.resize(n_A);
+
+        IdxVector xndvector;
+        // n_B or can we say that resize is constant 1?
+        xndvector.resize(n_B);
+        // n_B
+        IdxVector counter(n_B, 0);
+
```

Comment:
`n_B` (approximately; depending on STL version, compiler and distribution could be better (linux uses block heap memory, meaning it can choose to allocate e.g. 1kB with one allocation call if it is lucky and there's such a block free) or worse (Windows has contiguous memory so it has to look for a free spot) but i would just count `n_B` or maybe 2 or 3 times, but we're talking orders here - not exact values